### PR TITLE
feat(auth, web): migrate web to js_interop to be compatible with WASM

### PR DESCRIPTION
**Pull request by Lyokone**
## Description

*Replace this paragraph with a description of what this PR is doing. If you're modifying existing behavior, describe the existing behavior, how this PR is changing it, and what motivated the change.*

## Related Issues

#12027

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [ ] I read and followed the [Flutter Style Guide].
- [ ] I signed the [CLA].
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [X] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/firebase_auth/firebase_auth_web/lib/src/firebase_auth_web_user.dart [hunk 1]
```diff
@@ -31,12 +32,18 @@ class UserWeb extends UserPlatform {
                 isEmailVerified: _webUser.emailVerified,
                 isAnonymous: _webUser.isAnonymous,
                 creationTimestamp: _webUser.metadata.creationTime != null
-                    ? context['Date']
-                        .callMethod('parse', [_webUser.metadata.creationTime])
+                    ? (js_interop.globalContext.getProperty('Date'.toJS)!
+                            as js_interop.JSObject)
+                        .callMethod<js_interop.JSNumber>(
+                            'parse'.toJS, _webUser.metadata.lastSignInTime)
```
[the remark this answers is not in the document]
Comment (Lyokone) [reply packages/firebase_auth/firebase_auth_web/lib/src/firebase_auth_web_user.dart]: Fixed :)

**Review by exaby73 — APPROVED** [review]

File: packages/firebase_auth/firebase_auth_web/lib/src/interop/auth.dart [hunk 2]
```diff
@@ -123,43 +123,44 @@ class User extends UserInfo<auth_interop.UserJsImpl> {
...
-          .then(UserCredential.fromJsObject);
+      auth_interop.linkWithPopup(jsObject, provider.jsObject).toDart
+          as Future<UserCredential>;
 
   /// Links the authenticated [provider] to the user account using
   /// a full-page redirect flow.
-  Future<void> linkWithRedirect(AuthProvider provider) => handleThenable(
-      auth_interop.linkWithRedirect(jsObject, provider.jsObject));
+  Future<void> linkWithRedirect(AuthProvider provider) =>
+      auth_interop.linkWithRedirect(jsObject, provider.jsObject).toDart;
```
Comment (russellwheatley) [inline packages/firebase_auth/firebase_auth_web/lib/src/interop/auth.dart:156]:
```suggestion
  Future<void> linkWithRedirect(AuthProvider provider) =>
      auth_interop.linkWithRedirect(jsObject, provider.jsObject).toDart as Future<void>;
```
Is this correct?

File: packages/firebase_auth/firebase_auth_web/lib/src/interop/auth.dart [hunk 3]
```diff
@@ -168,26 +169,26 @@ class User extends UserInfo<auth_interop.UserJsImpl> {
...
-  Future<void> reauthenticateWithRedirect(AuthProvider provider) =>
-      handleThenable(
-          auth_interop.reauthenticateWithRedirect(jsObject, provider.jsObject));
+  Future<void> reauthenticateWithRedirect(AuthProvider provider) => auth_interop
+      .reauthenticateWithRedirect(jsObject, provider.jsObject)
+      .toDart;
 
   /// If signed in, it refreshes the current user.
-  Future<void> reload() => handleThenable(jsObject.reload());
+  Future<void> reload() => jsObject.reload().toDart;
```
Comment (russellwheatley) [inline packages/firebase_auth/firebase_auth_web/lib/src/interop/auth.dart:191]: I see `as Future<void>` isn't needed on other API with this type so perhaps it isn't needed like the others returning a type as opposed to void.

**Review by russellwheatley — APPROVED** [review]

File: packages/firebase_auth/firebase_auth_web/lib/src/interop/auth.dart (hunk 2 above)
Comment (Lyokone) [reply packages/firebase_auth/firebase_auth_web/lib/src/interop/auth.dart:156]: When you run .toDart, it is interpreted as Future<JSAny> which includes Future<void> so casting isn't necessary

File: packages/firebase_auth/firebase_auth_web/lib/src/utils/web_utils.dart [hunk 4]
```diff
@@ -4,38 +4,24 @@
...
-  if (_isFirebaseAuthError(e)) {
-    String? code = getProperty(e, 'code');
-    String? message = getProperty(e, 'message');
-    if (code == null || !code.startsWith('auth/')) return false;
-    if (message == null || !message.contains('Firebase')) return false;
+bool _hasFirebaseAuthErrorCodeAndMessage(JSError e) {
+  if (e.name == 'FirebaseError') {
+    String code = e.code ?? '';
+    String message = e.message ?? '';
+    if (code.startsWith('auth/')) return false;
```
Comment (JishnuM) [inline packages/firebase_auth/firebase_auth_web/lib/src/utils/web_utils.dart:23]: Are these checks correct? In the previous version of this, the checks were
```dart
    if (code == null || !code.startsWith('auth/')) return false;
    if (message == null || !message.contains('Firebase')) return false;
 ```

Now it is inverted to return false if the message contains `Firebase`, which seems wrong?

**Pull request merged**